Plutarch/TryFrom.hs

```diff
@@ -0,0 +1,404 @@
+        verhead = unTermCont $ fst <$> TermCont (ptryFrom @PData @(PAsData b) lsthead)
+    ver <-
+      tcont $
+        plet $
+          pdcons @s
+            # verhead
+            # pdnil
+    pure $ (punsafeCoerce opq, ver)
+
+class SumValidation (n :: Nat) (sum :: [[PLabeledType]]) where
+  validateSum :: Term s PData -> Term s (PBuiltinList PData)
+
+instance
+  {-# OVERLAPPABLE #-}
+  forall n x xs.
+  ( PTryFrom PData (PAsData (PDataRecord x))
+  , SumValidation (n + 1) xs
+  , KnownNat n
+  ) =>
+  SumValidation n (x ': xs)
+  where
+  validateSum s = unTermCont $
+    do
+      let n :: Integer
+          n = natVal (Proxy @n)
+      elem <- tcont $ plet $ pasConstr #$ s
+      let snd' :: Term _ (PBuiltinList PData)
+          snd' =
+            pif
+              (fromInteger n #== (pfstBuiltin # elem))
+              ( unTermCont $ do
+                  let rec = pdata $ psndBuiltin # elem
+                  y <- snd <$> TermCont (ptryFrom @PData @(PAsData (PDataRecord x)) $ pforgetData rec)
+                  pure $ punsafeCoerce (y :: Term _ (PTryFromExcess PData (PAsData (PDataRecord x))))
+              )
+              (validateSum @(n + 1) @xs $ punsafeCoerce s)
+      tcont $ plet snd'
+
+instance {-# OVERLAPPING #-} SumValidation n '[] where
+  validateSum _ = perror
+
+instance
+  {-# OVERLAPPING #-}
+  forall ys.
+  ( SumValidation 0 ys
+  ) =>
+  PTryFrom PData (PAsData (PDataSum ys))
+  where
+  type PTryFromExcess PData (PAsData (PDataSum ys)) = PUnit
+  ptryFrom opq = runTermCont $ do
+    _ <- tcont $ plet $ validateSum @0 @ys opq
+    pure (punsafeCoerce opq, pcon PUnit)
+
+----------------------- POpaque Instances -----------------------------------------------
+
+{- |
+    for none of the opaque instances it can be verified
+    that the actual structure is what it says to be
+    because that data is lost when the PAsData wrapper
+    is removed, this can only be safely used if you obtained
+    your POpaque safely
+-}
+instance
+  ( PTryFrom PData (PAsData a)
+  , PIsData a
+  ) =>
+  PTryFrom POpaque a
+  where
+  type PTryFromExcess POpaque a = PAsData a
+  ptryFrom opq = runTermCont $ do
+    let prop :: Term _ a
+        prop = punsafeCoerce opq
+    ver' <- fst <$> TermCont (ptryFrom @PData @(PAsData a) $ pforgetData $ pdata prop)
+    ver <- tcont $ plet ver'
+    pure $ (punsafeCoerce opq, ver)
+
+instance
+  ( PTryFrom a b
+  , PIsData a
+  , PIsData b
+  ) =>
+  PTryFrom (PAsData a) (PAsData b)
+  where
+  type PTryFromExcess (PAsData a) (PAsData b) = PTryFromExcess a b
+  ptryFrom opq = runTermCont $ do
+    ver' <- TermCont $ ptryFrom @a @b $ pfromData opq
+    ver <- tcont $ plet $ snd ver'
+    pure $ (punsafeCoerce opq, ver)
+
+----------------------- The class PFrom -------------------------------------------------
+
+{- |
+    Represents infallible conversion.
+    Laws:
+    - The result type must always carry less information than the origin
+      type
+-}
+class PFrom a b where
```

Comment:
I don't think we should have this honestly

Reply:
copied

Reply:
copied, I removed `PFrom` and `PMaybeFrom` for now